java-manual/modules/ROOT/pages/transactions.adoc

```diff
@@ -0,0 +1,487 @@
+        } else {
+            orgId = createOrganization(tx);
+            System.out.printf("Latest org is full, created %s.\n", orgId);
+            addPersonToOrganization(tx, name, orgId);
+            // If any of the above throws, the transaction will roll back
+            // -> not even Person is created!
+        }
+
+        return orgId;  // Organization ID to which the new Person ends up in
+    }
+
+    static String createOrganization(TransactionContext tx) {
+        var result = tx.run("""
+            CREATE (o:Organization {id: randomuuid(), createdDate: datetime()})
+            RETURN o.id AS id
+        """);
+        var org = result.single();
+        var orgId = org.get("id").asString();
+        return orgId;
+    }
+
+    static void addPersonToOrganization(TransactionContext tx, String personName, String orgId) {
+        tx.run("""
+            MATCH (o:Organization {id: $orgId})
+            MATCH (p:Person {name: $name})
+            MERGE (p)-[:WORKS_FOR]->(o)
+            """, Map.of("orgId", orgId, "name", personName)
+        );
+    }
+}
+----
+
+Should a transaction fail for a reason that the driver deems transient, it automatically retries to run the transaction function (with an exponentially increasing delay).
+For this reason, *transaction functions must be _idempotent_* (i.e., they should produce the same effect when run several times), because you do not know upfront how many times they are going to be executed.
+In practice, this means that you should not edit nor rely on globals, for example.
+Note that although transactions functions might be executed multiple times, the queries inside it will always run only once.
+
+A session can chain multiple transactions, but *only one single transaction can be active within a session at any given time*.
+To maintain multiple concurrent transactions, use multiple concurrent sessions.
+
+
+[#explicit-transactions]
+== Run an explicit transaction
+
+You can achieve full control over transactions by manually beginning one with the method `Session.beginTransaction()`, which returns a link:https://neo4j.com/docs/api/java-driver/current/org.neo4j.driver/org/neo4j/driver/Transaction.html[`Transaction`] object.
+You may then run queries inside an explicit transaction with the method `Transaction.run()`.
+
+[source, java]
+----
+try (var session = driver.session(SessionConfig.builder().withDatabase("neo4j").build())) {
+    try (Transaction tx = session.beginTransaction()) {
+        // use tx.run() to run queries
+        //     tx.commit() to commit the transaction
+        //     tx.rollback() to rollback the transaction
+    }
+}
+----
+
+An explicit transaction behaves as follows:
+
+- it gets committed automatically upon successful return, or manually through the method  link:https://neo4j.com/docs/api/java-driver/current/org.neo4j.driver/org/neo4j/driver/Transaction.html#commit()[`Transaction.commit()`]
```

Comment:
*This is not true.*

You *must* use `commit` prior to close.

See JavaDoc of transaction

```
Transactions are typically wrapped in a try-with-resources block
which ensures in case of any error in try block, the transaction is
automatically rolled back on close. Note that <code>ROLLBACK</code> is the
default action unless {@link #commit()} is called before closing.
```

Reply:
Ah, thanks for bringing this up. I was unsure but gave it for granted at this stage since this is how Python, Javascript, and Go work: they all default to commit if no error occurs. Interesting that Java took the opposite turn.

Reply:
Is this true? If so you, you should raise this with our colleagues. 

They will be in for a surprise wrt GQL: GQL requires exactly the behaviour of the Java driver, there will be no auto-commit.

Reply:
I checked the Python manual, and it does not specify that explicit transactions are automatically committed. And that's good, because they aren't.

This is what it says:
> Closing an explicit transaction can either happen automatically at the end of a with block, or can be explicitly controlled through the methods [Transaction.commit()](https://neo4j.com/docs/api/python-driver/current/api.html#neo4j.Transaction.commit), [Transaction.rollback()](https://neo4j.com/docs/api/python-driver/current/api.html#neo4j.Transaction.rollback), or [Transaction.close()](https://neo4j.com/docs/api/python-driver/current/api.html#neo4j.Transaction.close).

Admittedly, it's not super clear about what happens either.

Reply:
Weeeeell ... to me any return within a `with` block counts as reaching the end of the `with` block :grimacing:

Reply:
Either way, if the default behavior at the end of a `with` block is to commit, that's definitely different than in java, where the default behavior at the end of a `try` block (which is the same as `with` in python) is to rollback.

Reply:
No sorry. I might have expressed myself ambiguously. The Python driver will close transactions when leaving the with block. Closing an transaction that has not been committed or rolled back yet, means rolling it back.

TL;DR: explicit transactions are never committed unless the user does so explicitly (`tx.commit()`).

Reply:
Thanks for clarifying -- I misunderstood the driver's behavior 1.5 years ago, and now read in your message what I wanted to read :grimacing: I'll also clarify in other manuals then.